File: 7_graph/acwing/网络流/4-费用流/G - Dream Team.py
# https://zhuanlan.zhihu.com/p/496282947
# 题意
# 有n个人来自不同的学校ai,擅长不同的学科bi,每个人有一个能力值ci
# 要求组建一支i个人的梦之队最大化队员的能力值,并且满足队伍中所有人来自的学校和擅长的学科都不同.
# n<=3e4
# !ai,bi<=150 (暗示作为顶点的数据量)
# ci<=1e9

# 分析
# 把学校和学科看作点,把一个人看成匹配边,能力值看作边权,其实就是求有i条匹配边的最优匹配.可以用费用流解决.
# 此外题目要求输出匹配数为1,2,…k个匹配时的最优匹配.
# !在spfa费用流算法中一次spfa只会找到一条费用最小的增广流,
# !所以每次增广之后得到的费用就对应匹配数为1,2,…k个匹配时的答案.

# !能力值对应 流的费用
# !队伍里每个人学校和学科都不同:学校学科为虚拟源汇点，容量为1，这样就不会取到重复的学生了

import sys


sys.setrecursionlimit(int(1e9))
input = lambda: sys.stdin.readline().rstrip("\r\n")
MOD = int(1e9 + 7)
INF = int(1e18)

from collections import deque
from typing import List


class Edge:
    __slots__ = ("fromV", "toV", "cap", "cost", "flow")

    def __init__(self, fromV: int, toV: int, cap: int, cost: int, flow: int) -> None:
        self.fromV = fromV
        self.toV = toV
        self.cap = cap
        self.cost = cost
        self.flow = flow


class MinCostMaxFlow:
    """最小费用流的连续最短路算法复杂度为流量*最短路算法复杂度"""

    __slots__ = ("_n", "_start", "_end", "_edges", "_reGraph", "_dist", "_visited", "_curEdges")

    def __init__(self, n: int, start: int, end: int):
        """
        Args:
            n (int): 包含虚拟点在内的总点数
            start (int): (虚拟)源点
            end (int): (虚拟)汇点
        """
        assert 0 <= start < n and 0 <= end < n
        self._n = n
        self._start = start
        self._end = end
        self._edges: List["Edge"] = []
        self._reGraph: List[List[int]] = [[] for _ in range(n + 10)]  # 残量图存储的是边的下标

        self._dist = [INF] * (n + 10)
        self._visited = [False] * (n + 10)
        self._curEdges = [0] * (n + 10)

    def addEdge(self, fromV: int, toV: int, cap: int, cost: int) -> None:
        """原边索引为i 反向边索引为i^1"""
        self._edges.append(Edge(fromV, toV, cap, cost, 0))
        self._edges.append(Edge(toV, fromV, 0, -cost, 0))
        len_ = len(self._edges)
        self._reGraph[fromV].append(len_ - 2)
        self._reGraph[toV].append(len_ - 1)

    def work(self) -> List[int]:
        """
        Returns:
             List[int]: 每个k限流(匹配了k条边)的最大花费
        """
        res = []
        minCost = 0
        while self._spfa():
            # !限定流量为1 每次dfs只找到一条费用最小的增广流
            flow = self._dfs(self._start, self._end, 1)
            minCost += flow * self._dist[self._end]
            res.append(minCost)
        return res

    def _spfa(self) -> bool:
        """spfa沿着最短路寻找增广路径  有负cost的边不能用dijkstra"""
        n, start, end, edges, reGraph, visited = (
            self._n,
            self._start,
            self._end,
            self._edges,
            self._reGraph,
            self._visited,
        )

        self._curEdges = [0] * n
        self._dist = dist = [INF] * n
        dist[start] = 0
        queue = deque([start])

        while queue:
            cur = queue.popleft()
            visited[cur] = False
            for edgeIndex in reGraph[cur]:
                edge = edges[edgeIndex]
                cost, remain, next = edge.cost, edge.cap - edge.flow, edge.toV
                if remain > 0 and dist[cur] + cost < dist[next]:
                    dist[next] = dist[cur] + cost
                    if not visited[next]:
                        visited[next] = True
                        if queue and dist[queue[0]] > dist[next]:
                            queue.appendleft(next)
                        else:
                            queue.append(next)

        return dist[end] != INF

    def _dfs(self, cur: int, end: int, flow: int) -> int:
        if cur == end:
            return flow

        visited, reGraph, curEdges, edges, dist = (
            self._visited,
            self._reGraph,
            self._curEdges,
            self._edges,
            self._dist,
        )

        visited[cur] = True
        res = flow
        index = curEdges[cur]
        while res and index < len(reGraph[cur]):
            edgeIndex = reGraph[cur][index]
            next, remain = edges[edgeIndex].toV, edges[edgeIndex].cap - edges[edgeIndex].flow
            if remain > 0 and not visited[next] and dist[next] == dist[cur] + edges[edgeIndex].cost:
                delta = self._dfs(next, end, remain if remain < res else res)
                res -= delta
                edges[edgeIndex].flow += delta
                edges[edgeIndex ^ 1].flow -= delta
            curEdges[cur] += 1
            index = curEdges[cur]

        visited[cur] = False
        return flow - res


#####################################

n = int(input())
v = 150
START, END, OFFSET = 2 * v, 2 * v + 1, v
mcmf = MinCostMaxFlow(2 * v + 2, START, END)
for i in range(v):
    mcmf.addEdge(i + OFFSET, END, 1, 0)
    mcmf.addEdge(START, i, 1, 0)

for _ in range(n):
    u, v, w = map(int, input().split())
    u, v = u - 1, v - 1
    mcmf.addEdge(u, v + OFFSET, 1, -w)  # !要求最大费用流

res = mcmf.work()
print(len(res))
for i, cost in enumerate(res, 1):
    print(-cost)

File: 7_graph/网络流/4-费用流/G - Dream Team.go
// https://zhuanlan.zhihu.com/p/496282947
// 题意
// 有n个人来自不同的学校ai,擅长不同的学科bi,每个人有一个能力值ci
// 要求组建一支i个人的梦之队最大化队员的能力值,并且满足队伍中所有人来自的学校和擅长的学科都不同.
// 输出匹配数为1,2,…k个匹配时的最优匹配.
// n<=3e4
// !ai,bi<=150 (暗示作为顶点的数据量)
// ci<=1e9

// 分析
// 把学校和学科看作点,把一个人看成匹配边,能力值看作边权,
// 其实就是求有i条匹配边的最优匹配.可以用费用流解决.
// !对每个学生,从ai到bi连一条容量为1,费用为ci的边.
// 在spfa费用流算法中一次spfa只会找到一条费用最小的增广流,
// 所以每次增广之后得到的费用就对应匹配数为1,2,…k个匹配时的答案.

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n int
	fmt.Fscan(in, &n)
	schools := make([]int, n)
	types := make([]int, n)
	scores := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &schools[i], &types[i], &scores[i])
	}

	res := dreamTeam(schools, types, scores)
	fmt.Fprintln(out, len(res))
	for _, r := range res {
		fmt.Fprintln(out, r)
	}
}

func dreamTeam(schools []int, types []int, scores []int) (res []int) {
	n := len(schools)
	START, END, OFFSET := 320, 321, 160
	flow := NewMinCostMaxFlow(330, START, END)
	for i := 0; i < n; i++ {
		flow.AddEdge(schools[i], OFFSET+types[i], 1, -scores[i])
	}
	for i := 0; i < OFFSET; i++ {
		flow.AddEdge(START, i, 1, 0)
		flow.AddEdge(i+OFFSET, END, 1, 0)
	}

	curFlow := 0
	curCost := 0
	for {
		f, c := flow.FlowWithLimit(1) // !每次消耗1流量,每个回合的答案就是容量为1,2,...,时的最小费用
		if f == 0 {
			break
		}
		curFlow += 1
		curCost += -c
		res = append(res, curCost)
	}

	return
}

const INF int = 1e18

type Edge struct{ from, to, cap, flow, cost, id int }
type MinCostMaxFlow struct {
	AddEdge        func(from, to, cap, cost int)
	Flow           func() (maxFlow int, minCost int)
	FlowWithLimit  func(flowLimit int) (maxFlow int, minCost int)
	Slope          func() [][2]int // (flow, cost) 的每个转折点
	SlopeWithLimit func(flowLimit int) [][2]int
	Edges          func() []Edge // 注意根据from,to排除虚拟源点汇点; `flow>0` 才是流经的边
}

func NewMinCostMaxFlow(n, start, end int) *MinCostMaxFlow {
	type neighbor struct {
		to   int
		rid  int // rid 为反向边在邻接表中的下标
		cap  int // 边的残量
		cost int
		eid  int // -1表示是反向边
	}

	graph := make([][]neighbor, n)
	ei := 0
	addEdge := func(from, to, cap, cost, eid int) {
		graph[from] = append(graph[from], neighbor{to, len(graph[to]), cap, cost, eid})
		graph[to] = append(graph[to], neighbor{from, len(graph[from]) - 1, 0, -cost, -1})
	}

	dist := make([]int, len(graph))
	type vi struct{ v, i int }
	pre := make([]vi, len(graph))
	spfa := func() bool {
		for i := range dist {
			dist[i] = INF
		}
		dist[start] = 0
		inQueue := make([]bool, len(graph))
		inQueue[start] = true
		queue := []int{start}
		for len(queue) > 0 {
			cur := queue[0]
			queue = queue[1:]
			inQueue[cur] = false
			for i, edge := range graph[cur] {
				if edge.cap == 0 {
					continue
				}
				next := edge.to
				if cand := dist[cur] + int(edge.cost); cand < dist[next] {
					dist[next] = cand
					pre[next] = vi{cur, i}
					if !inQueue[next] {
						queue = append(queue, next)
						inQueue[next] = true
					}
				}
			}
		}
		return dist[end] < INF
	}

	// ek
	FlowWithLimit := func(flowLimit int) (maxFlow int, minCost int) {
		for maxFlow < flowLimit {
			if !spfa() {
				break
			}

			// 沿 st-end 的最短路尽量增广
			flow := INF
			for cur := end; cur != start; {
				p := pre[cur]
				if c := graph[p.v][p.i].cap; c < flow {
					flow = c
				}
				cur = p.v
			}
			for cur := end; cur != start; {
				p := pre[cur]
				edge := &graph[p.v][p.i]
				edge.cap -= flow
				graph[cur][edge.rid].cap += flow
				cur = p.v
			}
			maxFlow += flow
			minCost += dist[end] * flow
		}
		return
	}

	SlopeWithLimit := func(flowLimit int) (slope [][2]int) {
		maxFlow, minCost := 0, 0
		for maxFlow < flowLimit {
			if !spfa() {
				break
			}
			// 沿 st-end 的最短路尽量增广
			flow := INF
			for cur := end; cur != start; {
				p := pre[cur]
				if c := graph[p.v][p.i].cap; c < flow {
					flow = c
				}
				cur = p.v
			}
			for cur := end; cur != start; {
				p := pre[cur]
				edge := &graph[p.v][p.i]
				edge.cap -= flow
				graph[cur][edge.rid].cap += flow
				cur = p.v
			}
			maxFlow += flow
			minCost += dist[end] * flow
			slope = append(slope, [2]int{maxFlow, minCost})
		}
		return
	}

	AddEdge := func(from, to, cap, cost int) {
		addEdge(from, to, cap, cost, ei)
		ei++
	}

	GetEdges := func() (res []Edge) {
		for from, edges := range graph {
			for _, e := range edges {
				if e.eid == -1 {
					continue
				}
				res = append(res, Edge{from, e.to, e.cap + graph[e.to][e.rid].cap, graph[e.to][e.rid].cap, e.cost, e.eid})
			}
		}
		return
	}

	return &MinCostMaxFlow{
		AddEdge:        AddEdge,
		Flow:           func() (maxFlow int, minCost int) { return FlowWithLimit(INF) },
		FlowWithLimit:  FlowWithLimit,
		Slope:          func() [][2]int { return SlopeWithLimit(INF) },
		SlopeWithLimit: SlopeWithLimit,
		Edges:          GetEdges,
	}
}
